src/core.c/CompUnit/Repository/Distribution.pm6

```diff
@@ -54,6 +56,25 @@ class CompUnit::Repository::Distribution does Distribution {
         }
     }
 
+    # Try to locate a distribution by a given file name. Only makes sense for CURFS.
+    # $file is expected to be either absolute or relative to $*CWD.
+    method from-file(::?CLASS:U: $file, :$name, :$ver, :$auth, :$api --> ::?CLASS:D) is implementation-detail {
+        my @fs-repos = $*REPO.repo-chain.grep({ ($^repo ~~ CompUnit::Repository::Locally)
```

Comment:
Grepping for repo types like this is still a code smell

Reply:
What do you propose instead? The purpose of the method is to locate a distribution by a file belonging to it. So, it means

a. we talk about `IO::Path`, which is what `CUR::Locally` is
b. we don't consider installations because for them the distribution is known a priori

Reply:
You don't know what type of 3rd party repositories might be added externally by users later, hard coupling a bunch of this system when it shouldn't be.

Reply:
`CUR::Locally` is there for the need of `$repo.abspath` attribute.

So, to get rid of the filter there is a constraint is to be put on `:file` argument of `.candidates` (or rather one is to be removed): the path to file is not necessarily normalized into a form relative to the repository prefix. This would offload the task of normalization on repos themselves, but otherwise I would even be in favor of such change.

I'll try it.

Reply:
Addressed with 8006fdab0808c5ec7e73ca4a9c99b668d3c8a5e8

Reply:
What I mean is I still querying logic (i.e. calling `.candidates`) doesn't belong on a CUR-agnostic Distribution object. I would expect the call site (around `$Distribution.from-file(self.current_file()`) to make the call to `.candidates` or whatever.

Reply:
Yeah path normalization should be something repos do. That is -`:$file` should always be of the relative form that would be found in a META6.json field, and the repository is responsible for creatin

Reply:
Also long ago `CUR::Locally` was used for providing an `absolutify` method, and internally the code was called something like `my $abspath = $repo.?absolutify($path) // $path.absolute`, which seems like it was trying to provide an api for the path normalizaion / absolutification

Reply:
> What I mean is I still think querying logic (i.e. calling `.candidates`) doesn't belong on a CUR-agnostic Distribution object. I would expect the call site (around `$Distribution.from-file(self.current_file()`) to make the call to `.candidates` or whatever. `from-precomp` isn't querying so its not quite the same

I didn't know `Distribution` must be repo-agnostic. And, actually, it's wasn't full conforming to this constraint because it knows about the registry and expects a repository object for `CUR::Distribution`. Event the presence of `Repository` in its name makes it not so agnostic.

The other reason to have `from-file` in CURD was its similarity to `from-precomp`. Makes some sense to have both at the same location.

An alternative was to have the method planted onto `CU::RepositoryRegistry`. But then it breaks the semantic link between the two similar methods and would require longer self-explanatory name like `distribution-for-file`. Not that I would be against moving and renaming it, but the current implementation seemed like a bit more logical to me.

> That is -`:$file` should always be of the relative form that would be found in a META6.json field

In case of something like `raku -I$HOME/src/project $HOME/src/project/bin/script` CURFS should still be capable of resolving the full script name and locating it in meta.

> Also long ago `CUR::Locally` was used for providing an `absolutify` method

Normalization is slightly different as it tries to turn the file name into form usable for matching against meta keys.

To sum up things, I currently see conflicting requirements as to how `:file` must be treated. `run-script` uses the relative form of `"bin/$script"` to locate the installation repository. Yet, `raku -I. bin/script` would result in false positive on a `CURI` if the local directory distribution has been already installed earlier. So, it brings me back to the point where `from-file` does path normalization per-repo.

Reply:
I'm not sure path normalization such as changing `\` to `/` needs to happens when calling `file` -- the user should always be supplying the paths as they would be in a META6.json.  For example `"foo" : "lib\bar\baz.rakumod"` doesn't need to work, it should be `lib/bar/baz.rakumod`. It already can't be expected normalize things like casing without causing conflicts for itself, so in that regard its reasonable to just expect the user to a) only use prenormalized paths in their META6.json and b) only use those same style of paths when querying for a file. It *doesnt* really make sense for `:$file` to ever be an actual `IO::Path` object -- the more idiomatic name that is used elsewhere is `$name-path` which better describes it isn't meant to be a file object.

Reply:
> raku -I. bin/script

If the first repository, in this case `.` -- or `CURFS` of `$*CWD` -- contains a `bin/script` then it should always be chosen. Indeed it would never reach the point of asking `CURI` if it had that file. So I'm not sure Im understanding the issue.

Reply:
> I didn't know Distribution must be repo-agnostic. And, actually, it's wasn't full conforming to this constraint because it knows about the registry and expects a repository object for CUR::Distribution. Event the presence of Repository in its name makes it not so agnostic.

It used to be called `CU::Distribution if I remember correctly. I wasn't involved in naming that, but I did make the push for defining the `Distribution` interface, what it was meant to accomplish, and how it would do so. With that being said, maybe it doesn't conform but I'm not sure thats a reason to push it further in that direction. Querying repos from a Distribution object just seems antithetical to the `Distribution` `CompUnit::Repository` api and division of responsibility. I can mull over better solutions, but its going to take me more than a few days.

Reply:
> I'm not sure path normalization such as changing `\` to `/` needs to happens when calling `file` -- the user should always be supplying the paths as they would be in a META6.json.

I already have it locally with `candidates` doing nothing to `file` which is type-restricted to `Str:D`. Normalization is offloaded to `from-file` method where it does make more sense.

> So I'm not sure Im understanding the issue.

This is with the default course of events, I.e. when the repo chain is fully built by the core. But what if it has been customized by user and an installation repo of any kind happens to be included before CURFS variants?

> Querying repos from a Distribution object just seems antithetical

Ok, then `CURD.from-file` -> `CURR.distribution-for-file`. Actually, this would allow to make it a public interface, alongside with the rest of `*-for-*` methods.

Reply:
> such as changing `\` to `/`

Speaking of this, I can't quickly find if there is any requirement to always use `/` in META6. So, there gonna be a problem if a Win* user decides to go with `\` they are accustomed with.

But even more importantly, path normalization must, probably,  take care of converting the separator for a Win-platform.

Reply:
> Speaking of this, I can't quickly find if there is any requirement to always use / in META6.

I imagine it might work as-is,  but only coincidently and not in every scenario. The design, at least, is around the concept of the `name-path`, a relative-to-the-distribution-root path. We aren't concerned with how e.g. windows users might think they can write something that looks like a name-path but isn't, similarly to how we don't concern ourselves if users on case-insensitive systems might not capitalize name-paths appropriately -- i.e. even though on their windows system `lib/foo.rakumod` opens `lib/Foo.rakumod` for a CURFS,  only one of `$distribution.content("lib/Foo.rakumod")` or `$distribution.content("lib/foo.rakumod")` should open that file (the one that matches what is in the META6.json file, not all the variants that would happen to open a file on a given persons file system).

Reply:
> This is with the default course of events, I.e. when the repo chain is fully built by the core. But what if it has been customized by user and an installation repo of any kind happens to be included before CURFS variants?

I'm not sure you need to work around this. Lets say I `use Foo:ver<0.1>` in some script which gets loaded from some CURFS, and that `Foo` does a `use Foo::Submodule;`:  This PR wants that to effective be `use Foo::Submodule:ver<0.1>`, which shouldn't concern itself with which repository it is loading from (hence why we don't have load-from-this-repo-only stuff like `use Foo:from-repository<XXX>`).   Loading `Foo::Submodule:ver<0.1>` should be sufficient regardless of what repository it came from. At least theoretically.

Reply:
> I'm not sure you need to work around this.
> shouldn't concern itself with which repository it is loading from

I guess a concern might be that it could be strange to have two slightly different `$?DISTRIBUTION` objects floating about representing the same distribution. I'm not sure of how to describe a scenario where that is actually broken though.

Reply:
> Loading `Foo::Submodule:ver<0.1>` should be sufficient regardless of what repository it came from. At least theoretically.

You speak of module loading while my concern was about scripts and other kind of loading by path. Module loading logic is unaffected in this respect. 

But when there is a request for script `bin/script` specified on the command line, I barely see a user expecting it to come from an installation repository, no matter how it got into the head of the repo chain. Actually, I would expect the same to be true for `require "lib/Foo.rakumod"` where it is rather be expected a relative path, not a meta key.

And currently it all applies to `distribution-for-file` method only, whereas `candidates` (again) operate on keys only.